File: Solene.MobileApp/Solene.MobileApp.Core/Converters/QuestionJsonConverter.cs
using Newtonsoft.Json;
using Solene.Models;
using System;
using System.Collections.Generic;
using System.Text;

namespace Solene.MobileApp.Core.Converters
{
    public class QuestionJsonConverter : JsonConverter
    {
        public override bool CanRead => true;
        public override bool CanWrite => false;

        public override bool CanConvert(Type objectType)
        {
            return objectType == typeof(Question);
        }

        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
        {            
            if (reader.TokenType == JsonToken.StartObject)
            {
                existingValue = existingValue ?? serializer.ContractResolver.ResolveContract(objectType).DefaultCreator();
                serializer.Populate(reader, existingValue);
                if (existingValue is Question questionValue)
                {
                    questionValue.Text = questionValue.Text.Replace("\\n", Environment.NewLine);
                    return questionValue;
                }
                return existingValue;
            }
            else if (reader.TokenType == JsonToken.Null)
            {
                return null;
            }
            else
            {
                throw new JsonSerializationException();
            }
        }

        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
        {
            throw new NotImplementedException();
        }
    }
}

File: Solene.AdminClient.Crossplat/QuestionJsonConverter.fs
namespace Solene.AdminClient.Crossplat

open Newtonsoft.Json
open System
open Solene.Models
open Operators

type QuestionJsonConverter() = 
    inherit JsonConverter()

    override this.CanRead = true
    override this.CanWrite = false

    override this.CanConvert (objectType: Type) : bool =        
        objectType = typeof<AdminQuestion>

    override this.ReadJson (reader: JsonReader, objectType: Type, existingValue: obj, serializer: JsonSerializer) : obj =
        match reader.TokenType with
        | JsonToken.StartObject -> 
            let value = 
                if existingValue = null then null 
                else serializer.ContractResolver.ResolveContract(objectType).DefaultCreator
            match value with
            | null -> null
            | _ -> 
                serializer.Populate(reader, existingValue)
                match existingValue with 
                | t when t.GetType() = typeof<AdminQuestion> -> 
                    let question = existingValue :?> AdminQuestion
                    question.Text <- question.Text.Replace("\\n", Environment.NewLine)
                    question :> obj
                | _ -> existingValue
        | JsonToken.Null ->
            null
        | _ -> raise (new JsonSerializationException())

    override this.WriteJson (writer: JsonWriter, value: obj, serializer: JsonSerializer) : unit =
        raise (new NotImplementedException())
        

